### PR TITLE
add ErrorReason/ErrorStatus to CRD statuses

### LINKED ISSUE #97: Add ErrorReason/ErrorStatus to CRD Statuses
/kind feature

**Describe the solution you'd like**
our configs can end up in an error state, such as having a cluster with two configs that both only have join & init configuration. In this case we need to be able to report up that one is broken and needs attention.

In order to do this we will need ErrorReason and ErrorStatus on our CRDs.

**Anything else you would like to add:**
Look here but do not use the same capierrors.MachineStatusError type since that doesn't apply to this case.

https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha2/machine_types.go#L111

Also please regenerate CRDs after the new fields are added.

/priority important-longterm
/milestone v0.1
^ note on milestone, it can slip due to lower priority

### PR DESCRIPTION
**Pull request by mytunguyen**
**What this PR does / why we need it**:
Add ErrorReason/ErrorStatus to CRD statuses so it's easier to debug. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #97 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by vincepri** [comment]
LGTM

/ok-to-test

File: controllers/kubeadmconfig_controller.go [hunk 1]
```diff
@@ -95,6 +96,8 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	machine, err := util.GetOwnerMachine(ctx, r.Client, config.ObjectMeta)
 	if err != nil {
 		log.Error(err, "could not get owner machine")
+                config.Status.ErrorStatus = CreateKubeadmconfigError
```
Comment (detiber) [inline controllers/kubeadmconfig_controller.go]: Should this be a non-retryable error? I would probably expect this to be a transient error only.

We should only set ErrorMessage and ErrorReason for non-retryable errors.

Comment (vincepri) [reply controllers/kubeadmconfig_controller.go]: Agreed, I missed this change when I first glanced through the PR.

@mytunguyen ErrorStatus and ErrorReason are usually only set in failures, when you usually require user intervention for example.

File: controllers/kubeadmconfig_controller.go [hunk 2]
```diff
@@ -95,6 +96,8 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
...
+                config.Status.ErrorReason = "could not get owner machine"
 		return ctrl.Result{}, err
```
Comment (chuckha) [inline controllers/kubeadmconfig_controller.go]: should this even be returning err?

Comment (ncdc) [reply controllers/kubeadmconfig_controller.go]: No, I don't think it should. We should wait for the CAPI machine controller to set the ownerReference in the KubeadmConfig to the Machine. Once that happens, this controller will get a natural update event and rereconcile the KubeadmConfig.

Comment (detiber) [reply controllers/kubeadmconfig_controller.go]: Agreed, no need to return an error here.

File: controllers/kubeadmconfig_controller.go (hunk 1 above)
Comment (mytunguyen) [reply controllers/kubeadmconfig_controller.go]: makes sense. okay!

File: controllers/kubeadmconfig_controller.go (hunk 2 above)
Comment (mytunguyen) [reply controllers/kubeadmconfig_controller.go]: all set I think

**Comment by vincepri** [comment]
/assign @chuckha 

for approval

**Comment by chuckha** [comment]
@vincepri What do you think about aligning this with the Machine and Cluster type from CAPI and using `ErrorReason` && `ErrorMessage`? I'm still trying to figure out if `ErrorReason` even makes sense here 🤔  (the machine readable error message. what would actually be consuming this error message? i think this is only for humans?)

**Comment by vincepri** [comment]
+1 for consistency, the plan sounds good to me. I'd keep both for now, until we actually have a large failure-mode handling plan

**Comment by chuckha** [comment]
/approve

Thanks for adding these!

**Comment by chuckha** [comment]
oh, can you squash your commits?

**Comment by mytunguyen** [comment]
Sweet! Did it

**Comment by vincepri** [comment]
Good timing :D

**Pull request merged**